### PR TITLE
Python: Smithy-Dafny's StandardLibrary

### PR DESCRIPTION
**Pull request by lucasmcdonald3**
*Issue #, if available:*

*Description of changes:*

* Commit StandardLibrary changes for Python (and Dafny 4.7)
  * When I merge Python into main, if there are conflicts between Python's 4.7 upgrade and main's (NET/Java) 4.7 upgrade, I plan to take in main's.
* Apply an `@protocolDefinition` trait to localService. "Real" Smithy code generators generate code for protocols. (A canonical protocol is HTTP.) Polymorph-NET/Java are not "real" Smithy code generators and instead hardcode localService. However,  Polymorph Python extends a "real" Smithy code generator (Smithy-Python). Applying this trait lets "real" Smithy code generators identify the localService trait as a valid protocol.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: TestModels/dafny-dependencies/Model/traits.smithy [hunk 1]
```diff
@@ -24,6 +24,7 @@ list ServiceList {
 // A trait for explicitly modeling the configuration options that should be
 // available in the generated methods for creating clients.
 @trait(selector: "service")
+@protocolDefinition
```
Comment (josecorella) [inline TestModels/dafny-dependencies/Model/traits.smithy:27]: oooh this is new, at least for me, can we add in the pr description what this is for and why it is being added?

File: TestModels/dafny-dependencies/StandardLibrary/Makefile [hunk 2]
```diff
@@ -78,21 +88,19 @@ transpile_dependencies:
...
-		-spillTargetCode:3 \
-		-compile:0 \
-		-optimizeErasableDatatypeWrapper:0 \
-		-compileSuffix:1 \
-		-unicodeChar:0 \
-		-functionSyntax:3 \
-		-useRuntimeLib \
-		$(DAFNY_OPTIONS) \
-		-out $(OUT)
+	translate $(TARGET) \
```
Comment (josecorella) [inline TestModels/dafny-dependencies/StandardLibrary/Makefile:94]: no `$(DANFY_OPTIONS)`?

File: TestModels/dafny-dependencies/StandardLibrary/runtimes/python/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+packages = [
+    { include = "standard_library", from = "src" },
+]
+# Include all of the following .gitignored files in package distributions,
+# even though it is not included in version control
+include = ["**/internaldafny/generated/*.py"]
+
+[tool.poetry.dependencies]
+python = "^3.11.0"
+DafnyRuntimePython = "^4.4.0"
```
Comment (josecorella) [inline TestModels/dafny-dependencies/StandardLibrary/runtimes/python/pyproject.toml]: ^4.7.0?

File: TestModels/dafny-dependencies/StandardLibrary/runtimes/python/src/standard_library/internaldafny/extern/UTF8.py
Comment (josecorella) [inline TestModels/dafny-dependencies/StandardLibrary/runtimes/python/src/standard_library/internaldafny/extern/UTF8.py:1]: nice comments! made it very easy to follow

File: TestModels/dafny-dependencies/Model/traits.smithy (hunk 1 above)
Comment (lucasmcdonald3) [reply TestModels/dafny-dependencies/Model/traits.smithy:27]: Done

File: TestModels/dafny-dependencies/StandardLibrary/Makefile (hunk 2 above)
Comment (lucasmcdonald3) [reply TestModels/dafny-dependencies/StandardLibrary/Makefile:94]: Added

File: TestModels/dafny-dependencies/StandardLibrary/runtimes/python/pyproject.toml (hunk 3 above)
Comment (lucasmcdonald3) [reply TestModels/dafny-dependencies/StandardLibrary/runtimes/python/pyproject.toml]: Sure! This is now possible 🎉

**Review by josecorella — APPROVED** [review]

**Pull request merged**